src/Features/Core/Portable/ExternalAccess/VSTypeScript/Api/VSTypeScriptCompletionServiceWithProviders.cs

```diff
@@ -22,25 +23,94 @@ internal sealed override CompletionRules GetRules(CompletionOptions options)
+
+            try
+            {
+                tsMatchResults.AddRange(matchResults.Where(static r => r.ShouldBeConsideredMatchingFilterText)
+                    .Select(static r => new VSTypeScriptCompletionItemMatchResult(r)));
+
+                FilterItemsImpl(document, tsMatchResults, filterText, tsFilteredMatchResults);
+                builder.AddRange(tsMatchResults.Select(static r => r.MatchResult));
+            }
+            finally
+            {
+                // Don't call ClearAndFree, which resets the capacity to a default value.
+                tsMatchResults.Clear();
+                tsFilteredMatchResults.Clear();
+                s_listOfTSMatchResultPool.Free(tsMatchResults);
+                s_listOfTSMatchResultPool.Free(tsFilteredMatchResults);
+            }
+        }
 
+        // Override this to bypass the expensive call to public FilterItems method.
         internal virtual void FilterItemsImpl(
             Document document,
-            IReadOnlyList<MatchResult> matchResults,
+            IReadOnlyList<VSTypeScriptCompletionItemMatchResult> tsMatchResults,
             string filterText,
-            IList<MatchResult> builder)
+            IList<VSTypeScriptCompletionItemMatchResult> builder)
         {
+            using var _1 = ArrayBuilder<CompletionItem>.GetInstance(tsMatchResults.Count, out var itemsBuilder);
+            using var _2 = PooledDictionary<CompletionItem, VSTypeScriptCompletionItemMatchResult>.GetInstance(out var map);
+
+            foreach (var result in tsMatchResults)
+            {
+                itemsBuilder.Add(result.CompletionItem);
+                map.Add(result.CompletionItem, result);
+            }
+
 #pragma warning disable RS0030 // Do not used banned APIs
-            var filteredItems = FilterItems(document, matchResults.SelectAsArray(item => item.CompletionItem), filterText);
+            var filteredItems = FilterItems(document, itemsBuilder.ToImmutable(), filterText);
 #pragma warning restore RS0030 // Do not used banned APIs
 
             var helper = CompletionHelper.GetHelper(document);
-            builder.AddRange(filteredItems.Select(item => helper.GetMatchResult(item, filterText, includeMatchSpans: false, CultureInfo.CurrentCulture)));
+            builder.AddRange(filteredItems.Select(item => map[item]));
+        }
+
+        internal static void FilterItemsImplDefault(
+            Document document,
+            IReadOnlyList<VSTypeScriptCompletionItemMatchResult> tsMatchResults,
+            string filterText,
+            IList<VSTypeScriptCompletionItemMatchResult> tsBuilder)
+        {
+            var matchResults = s_listOfMatchResultPool.Allocate();
+            var builder = s_listOfMatchResultPool.Allocate();
+
+            try
+            {
+                matchResults.AddRange(tsMatchResults.Select(static r => r.MatchResult));
+                FilterItemsDefault(CompletionHelper.GetHelper(document), matchResults, filterText, builder);
```

Comment:
might be helpful to fully qualify references to FilterItemsDfault to indicate where it is.